resources/common.js

```diff
@@ -80,3 +80,157 @@ let Util = {
+            });
+
+            ws.addEventListener('open', () => {
+                console.log('Connected!');
+
+                ws.send(JSON.stringify({
+                    call: 'subscribe',
+                    events: Object.keys(subscribers),
+                }));
+
+                for (let msg of queue) {
+                    ws.send(JSON.stringify(msg));
+                }
+
+                queue = null;
+            });
+
+            ws.addEventListener('message', (msg) => {
+                try {
+                    msg = JSON.parse(msg.data);
+                } catch(e) {
+                    console.error('Invalid message received: ', msg);
+                    return;
+                }
+
+                if (msg.rseq !== undefined && responsePromises[msg.rseq]) {
+                    responsePromises[msg.rseq](msg);
+                    delete responsePromises[msg.rseq];
+                } else {
+                    processEvent(msg);
+                }
+            });
+
+            ws.addEventListener('close', () => {
+                queue = [];
+
+                console.log('Trying to reconnect...');
+                // Don't spam the server with retries.
+                setTimeout(() => {
+                    connectWs();
+                }, 300);
+            });
+        }
+
+        connectWs();
+    } else {
+        function waitForApi() {
+            if (!window.OverlayPluginApi || !window.OverlayPluginApi.ready) {
+                setTimeout(waitForApi, 300);
+                return;
+            }
+
+            let q = queue;
+            queue = null;
+
+            window.__OverlayCallback = processEvent;
+
+            OverlayPluginApi.callHandler(JSON.stringify({
+                'call': 'subscribe',
+                'events': Object.keys(subscribers),
+            }), null);
+
+            for (let [msg, resolve] of q) {
+                OverlayPluginApi.callHandler(JSON.stringify(msg), resolve);
+            }
+        }
+
+        waitForApi();
+    }
+
+    function processEvent(msg) {
+        if (subscribers[msg.type]) {
+            for (let sub of subscribers[msg.type]) {
+                sub(msg);
+            }
+        } else if (msg.type !== 'CombatData') {
+            console.warn(`Received event ${msg.type} even though we're not subscribed!`);
+        }
+    }
+
+    function sendMessage(msg, cb) {
+        if (ws) {
+            if (queue) {
+                queue.push(msg);
+            } else {
+                ws.send(JSON.stringify(msg));
+            }
+        } else {
+            if (queue) {
+                queue.push([msg, cb]);
+            } else {
+                OverlayPluginApi.callHandler(JSON.stringify(msg), cb);
+            }
+        }
+    }
+
+    window.addOverlayListener = (event, cb) => {
+        if (!subscribers[event]) {
+            subscribers[event] = [];
+
+            if (!queue) {
+                sendMessage({
+                    call: 'subscribe',
```

Comment:
Yes. Each subscription (regardless of whether it happens over WebSocket or the `OverlayPluginApi`) is just a call to [`EventDispatcher.Subscribe()`](https://github.com/ngld/OverlayPlugin/blob/master/OverlayPlugin.Core/EventDispatcher.cs). That class handles all events and handler calls. Since both work the same from overlay and WebSocket connections, I thought abstracting the actual implementation away from the transport would be the easiest solution.
Also, having an event filter seemed necessary to me for WebSocket support since sending all supported events to a pure DPS overlay would result in way too much wasted processing time.

DPS data is sent by the `MiniParseEventSouce` and for now it's just limited to once per second though I could expose that limit through the UI and make it configurable. However, that rate would always apply to all overlays due to the way events work right now.

Reply:
Hmm, ok.  For what it's worth, cactbot also sends dps info from FightTracker.  So, in a new world where`MiniParseEventSource` is sending the fight info, that code and event probably need to be deleted?

Yeah, an option on `MiniParseEventSource` to rate limit dps would be great, even if that applies to all things using that info.  I've seen that chew up a lot of cpu in alliance raids or open worlds.  (There's a lot of combatants and a lot of unnecessary fields in it.  Maybe it'd be worth filtering some out.  For example, https://pastebin.com/TuRK7xrR)

Reply:
Filtering sounds like a good idea but is hard to do in practice since you never know which overlay relies on which fields.
I'll have to check with the recent changes but the last times I checked, the whole thing took usually less than 100 ms which seems perfectly fine to me.
One issue might be that it spawns a lot of tasks (1 + number of combatants in ACT's main table) each time a new update is generated. If that's the issue, limiting the amount of tasks it can spawn should solve the issue and is easy (since it's just a parameter to `Parallel.ForEach`).